lib/improver/tests/utilities/test_cube_extraction.py

```diff
@@ -0,0 +1,165 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Unit tests for cube extraction utilities """
+
+import unittest
+import os
+import numpy as np
+from tempfile import mkdtemp
+from subprocess import call
+
+import iris
+from iris.tests import IrisTest
+from iris.coords import DimCoord
+from iris.exceptions import CoordinateNotFoundError
+
+from improver.utilities.cube_extraction import (parse_constraint_list,
+                                                extract_subcube)
+
+
+def set_up_precip_probability_cube():
+    """
+    Set up a cube with spatial probabilities of precipitation at three
+    exceedance thresholds
+    """
+    data = np.array([[[0.85, 0.95, 0.73],
+                      [0.75, 0.85, 0.65],
+                      [0.70, 0.80, 0.62]],
+                     [[0.18, 0.20, 0.15],
+                      [0.11, 0.16, 0.09],
+                      [0.10, 0.14, 0.03]],
+                     [[0.03, 0.04, 0.01],
+                      [0.02, 0.02, 0.00],
+                      [0.01, 0.00, 0.00]]])
+
+    mmh_to_ms = 0.001/3600.
+    threshold = DimCoord(mmh_to_ms*np.array([0.03, 0.1, 1.0]),
+                         long_name="threshold", units="m s-1")
+    ycoord = DimCoord(np.arange(3), "projection_y_coordinate")
+    xcoord = DimCoord(np.arange(3), "projection_x_coordinate")
+
+    cube = iris.cube.Cube(data, long_name="probability_of_precipitation",
+                          dim_coords_and_dims=[(threshold, 0), (ycoord, 1),
+                                               (xcoord, 2)], units="1")
+    return cube
+
+
+class Test_parse_constraint_list(IrisTest):
+
+    """ Test function to parse constraints and units into dictionaries """
+
+    def setUp(self):
+        """ Set up some constraints to parse """
+        self.constraints = ["percentile=10", "threshold=0.1"]
+
+    def test_basic_no_units(self):
+        """ Test simple key-value splitting with no units """
+        cdict, udict = parse_constraint_list(self.constraints, None)
+        self.assertEqual(cdict["percentile"], 10)
+        self.assertEqual(cdict["threshold"], 0.1)
+        self.assertFalse(udict)
+
+    def test_some_units(self):
+        """ Test units list containing "None" elements is correctly parsed """
+        units = ["None", "mm h-1"]
+        _, udict = parse_constraint_list(self.constraints, units)
+        self.assertEqual(udict["threshold"], "mm h-1")
+        self.assertNotIn("percentile", udict.keys())
+
+    def test_unmatched_units(self):
+        """ Test for ValueError if units list does not match constraints """
+        units = ["mm h-1"]
+        msg = "units list must match constraints"
+        with self.assertRaisesRegexp(ValueError, msg):
+            parse_constraint_list(self.constraints, units)
+
+
+class Test_extract_subcube(IrisTest):
+
+    """ Test function to extract subcube according to constraints """
+
+    def setUp(self):
```

Comment:
Unless there is a good reason, I don't think we should be creating files/directories and saving the output cube.

Reply:
If this is a reference to the [mock module](https://docs.python.org/3/library/unittest.mock.html), which has been incorporated into the unittest module in Python 3, then we should probably have a think about whether we want to start using this functionality, or whether it adds unnecessary complexity.

Reply:
I've got rid of the need for this now, by loading the cube in the CLI and passing that into the extract_subcube() function.  This more closely mirrors the convention throughout the rest of the IMPROVER code.